### DISCUSSION TITLE
Can HypergraphConv used for directed Hypergraph

### DISCUSSION THREAD

**Discussion by CaiYitao** [pyg-team/pytorch_geometric#9183] [General]
we have a chemical reaction network, which can be represented as a derivation graph/ directed hypergraph. I am trying to use HypergraphConv class to do convolution on it. But it seems not working. 
Please see attached hypergraph. the r_i are hyperedges, molecules are nodes. and the hyperedges indicate reaction rules/types which connect reactants to products. 

here is my code 
```python
from torch_geometric.nn import HypergraphConv

class HyperGraphLayer(nn.Module):
    def __init__(self, cfg):
        super().__init__()


        self.dim_h = cfg.dim_h
        self.num_heads = cfg.num_heads
        self.attn_dropout = cfg.attn_dropout
        self.layer_norm = cfg.layer_norm
        self.batch_norm = cfg.batch_norm
        # self.lin = nn.Linear(cfg.dim_h * cfg.hg_attn_heads, cfg.dim_h)
        self.hypergraph_conv = HypergraphConv(cfg.dim_h, cfg.dim_h, use_attention=True)
        self.attn = nn.MultiheadAttention(cfg.dim_h, cfg.num_heads, dropout=self.attn_dropout, batch_first=True)

        self.mlp = nn.Sequential(
            nn.Linear(cfg.dim_h, cfg.dim_h*2),
            nn.GELU(),
            nn.Dropout(cfg.dropout),
            nn.Linear(cfg.dim_h*2, cfg.dim_h),
            nn.GELU(),
            nn.Dropout(cfg.dropout)
        )

        self.layer_norm_local = nn.LayerNorm(cfg.dim_h)
        self.layer_norm_attn = nn.LayerNorm(cfg.dim_h)
        self.layer_norm = nn.LayerNorm(cfg.dim_h)
        
    def forward(self, hypergraph):
        h = hypergraph.x
        h_in = h

        print("hyperedge_index: ", hypergraph.edge_index, "hyperedge_attr shape: ", hypergraph.edge_attr.shape)
        print("hyperedge_index[1]", hypergraph.edge_index[1] )
     
        h_local = self.hypergraph_conv(h, hypergraph.edge_index, hyperedge_attr = hypergraph.edge_attr)
       
        h_local = h_local + h_in
        h_local = self.layer_norm_local(h_local)
        h_attn = self.attn(h, h, h, need_weights=False)[0]
   
        h_attn = h_attn + h_in
        h_attn = self.layer_norm_attn(h_attn)

        h= h_local + h_attn
        h = h + self.mlp(h)
        h = self.layer_norm(h)

        hypergraph.x = h
        return hypergraph

```
but it gives error:
```
hyperedge_index: tensor([[5, 1, 4, 3, 2, 6],
        [0, 0, 0, 1, 1, 1]]) 
hyperedge_index[1]: tensor([0, 0, 0, 1, 1, 1])
hyperedge_attr shape:  torch.Size([1, 128])
Traceback (most recent call last):
  
    h_local = self.hypergraph_conv(h, hypergraph.edge_index, hyperedge_attr = hypergraph.edge_attr)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
    x_j = hyperedge_attr[hyperedge_index[1]]
          ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^
IndexError: index 1 is out of bounds for dimension 0 with size 1
```
since currently the hypergraph only have one hyperedge which from reactants set mol 5, mol1, mol4  to products set mol3, mol2, mol6.  the hypergraph conv seems can not handle this case. 
I am not sure if I understand correctly about your HypergraphConv document, I assume that hyperedge_index: tensor([[5, 1, 4, 3, 2, 6],
        [0, 0, 0, 1, 1, 1]])     means that the nodes 5, 1 ,4 connected with 3 ,2 ,6 by a directed hyperedge.
Looking forward to your suggestions. 

[summary_2.pdf](https://github.com/pyg-team/pytorch_geometric/files/14931831/summary_2.pdf)

**Comment by rusty1s** [comment]
Sorry for late reply. Why does your `hyperedge_attr` only holds features for a single hyperedge? Changing this should resolve your issue.

**Reply by CaiYitao** [reply]
thank you for your kind reply.   in the printout case, the hypergraph now only have one hyperedge, our hypergraph is directed hypergraph.  can the HypergraphConv used for directed one?  from the paper in the doc suggested, it seems not the case. did I misunderstood anything?

**Reply by rusty1s** [reply]
What do you mean by directed hypergraph? Hypergraphs are always directed, right?

**Reply by CaiYitao** [reply]
It seems not the case, directed hypergraph have two incidence matrix one is source(tail) nodes to hyperedge and the other is hyperedge to target(head) nodes
![image](https://github.com/pyg-team/pytorch_geometric/assets/65641330/5f078323-a304-4a0f-aa0c-6dd552b6d0b9)

I find the following papers are relevant. 
You can refer to these papers.

[IET Intelligent Trans Sys - 2021 - Luo - Directed hypergraph attention network for traffic forecasting.pdf](https://github.com/pyg-team/pytorch_geometric/files/14993827/IET.Intelligent.Trans.Sys.-.2021.-.Luo.-.Directed.hypergraph.attention.network.for.traffic.forecasting.pdf)
[heterogenousHyperGraphNN.pdf](https://github.com/pyg-team/pytorch_geometric/files/14993829/heterogenousHyperGraphNN.pdf)


[Directed_Hypergraph_Neural_Network.pdf](https://github.com/pyg-team/pytorch_geometric/files/14993831/Directed_Hypergraph_Neural_Network.pdf)

**Comment by CaiYitao** [comment]
Could you please explain the concept of `hyperedge_index` in the context of the `HypergraphConv` class? Here's my understanding:

For example, I believe `hyperedge_index: tensor([[5, 1, 4, 3, 2, 6, 7, 9, 8, 11, 12], [0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1]])` represents directed hyperedges in a hypergraph.  It seems to indicate the following:

* **Hyperedge 1:** Source nodes (5, 1, 4) are connected to target nodes (3, 2, 6).
* **Hyperedge 2:** Source nodes (7, 9) are connected to target nodes (8, 11, 12).

The 0s in the second row denote source nodes, and 1s denote target nodes.  Transitions from 1 to 0 likely mark the beginning of new hyperedges.

Could you please confirm if this understanding is in the same page as in the HypergraphConv class?

**Reply by rusty1s** [reply]
`hyperedge_index` should represent the non-zero indices of the incidence matrix of a hypergraph. In your case,
```
hyperedge_index: tensor([[5, 1, 4, 3, 2, 6, 7, 9, 8, 11, 12], [0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1]])
```
this means that `[5, 1, 4, 7, 9]` define a hyperedge, and `[3, 2, 6, 8, 11, 12]` define a hyperedge (i.e., all indices in `hyperedge_index[0]` that share an index in `hyperedge_index[1]`.

**Reply by CaiYitao** [reply]
Thank you very much! so if we need to construct a new directed hypergraph conv class, the hyperedge index of this example should be 
`hyperedge_index: tensor([[5, 1, 4, 3, 2, 6, 7, 9, 8, 11, 12], [0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1]]) ` and then we build an extra index indicate the source to target, `idx_source_targe: tensor([0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1])` 0s indicate source/tail, 1s indicate target/head of each hyperedge, from 1 to 0 indicates another hyperedge. 

### the following is a real case 
 after I have done such changes, 
`hyperedge_index after transform: tensor([[5, 6, 7],
        [0, 0, 0]])  `  indicate there is only one hyperedge connect with 5,6,7. and `idx_source_target: tensor([0, 1, 1])`,
  the node indices 5, 6, 7 came from a derivation graph which indexed by [0, numVertices + numEdges]. 
      There is an error in index_select for pytorch geometric. 
 ``` File "/.cache/pyg/message_passing/model.layer_DirectedHGConv_edge_updater.py", line 103, in edge_collect
    edge_attr_i = self._index_select(edge_attr, edge_index_i)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/lib/python3.11/site-packages/torch_geometric/nn/conv/message_passing.py", line 300, in _index_select
    return self._index_select_safe(src, index)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/lib/python3.11/site-packages/torch_geometric/nn/conv/message_passing.py", line 315, in _index_select_safe
    raise IndexError(
IndexError: Found indices in 'edge_index' that are larger than 0 (got 7). Please ensure that all indices in 'edge_index' point to valid indices in the interval [0, 1) in your node feature matrix and try again.
```
    


if we have a hyperpergraph which does not have hyperedges ,only the individual vertices, at the beginning, and at some time point there are hyperedges appeared, so the vertices indices within hyperedges could happen to have [ 5, 6, 7 ] , or something like[10, 28, 62,...] that.  Does it mean that we should reindex the vertices within hyperedges to be selfcontained in hyperedges? and I do not understand why we should `ensure that all indices in 'edge_index' point to valid indices in the interval [0, 1) in your node feature matrix`

**Comment by CaiYitao** [comment]
It is actually related to this issue: https://github.com/qiankunmu/HDHGN/issues/3
